swift/downgrades/04ef9ecaa9e06a587f252f147462861e3d32846a/downgrade.ql

```diff
@@ -0,0 +1,47 @@
+class Element extends @element {
+  string toString() { none() }
+}
+
+query predicate unspecified_elements(Element e, string property, string error) {
+  unspecified_elements(e, property, error)
+  or
+  error = "Removed parameter packs during the database downgrade" and
+  property = "" and
+  (
+    pack_element_exprs(e, _) or
+    pack_expansion_exprs(e, _) or
+    pack_element_types(e, _) or
+    pack_expansion_types(e, _, _) or
+    pack_types(e) or
+    element_archetype_types(e)
+  )
+}
+
+query predicate unspecified_element_children(Element e, int index, Element child) {
+  unspecified_element_children(e, index, child)
+  or
+  pack_element_exprs(e, child) and index = 0
+  or
+  pack_expansion_exprs(e, child) and index = 0
+  or
+  pack_element_types(e, child) and index = 0
+  or
+  (
+    pack_expansion_types(e, child, _) and index = 0
+    or
+    pack_expansion_types(e, _, child) and index = 1
+  )
+  or
+  pack_type_elements(e, index, child)
+}
+
+query predicate unspecified_element_parents(Element e, Element parent) {
+  unspecified_element_parents(e, parent) or
+  pack_element_exprs(parent, e) or
+  pack_expansion_exprs(parent, e) or
+  pack_element_types(parent, e) or
+  pack_expansion_types(parent, e, _) or
+  pack_expansion_types(parent, _, e) or
+  pack_types(parent, e) or
+  element_archetype_types(parent, e)
+}
```

Comment:
again, we don't have an established parent/child relationship for types.

For `expr`s, this would be incomplete and wrong, because a removed pack expression could be the child of a non-pack expression, and in turn the child of a pack expression could be a non-pack expression. However I think it's ok if we don't change this relation at all: its main objective is to record where an extractor-unspecified entity occurs (that is, where the Swift compiler gives us a `nullptr`). For upgrade/downgrade unspecified elements, I think it's ok to not have that information recorded in the DB directly. It's still recorded indirectly in the parent itself.